demo/cpp/main.cpp

```diff
@@ -0,0 +1,288 @@
+    if( !context ) {
+		std::cout << "failed to create execution context" << std::endl;
+		return 1;
+	}
+
+    //  Specify (fix) model input layer size
+    //  fill free to try different input layer size in range used while model convertion
+    cv::Size input_layer_size(960, 544);
+    context->setBindingDimensions(0, nvinfer1::Dims4(1, 3, input_layer_size.height, input_layer_size.width));
+    
+    //  Create image transformer
+    ImageTransformer image_transformer(input_layer_size);
+    std::cout << "The model has been successfully parsed. Input dims set." << std::endl;
+
+    //  Allocate input/output memory
+    std::vector< nvinfer1::Dims > input_dims;   // we expect only one input
+    std::vector< nvinfer1::Dims > output_dims;  // and one output
+    std::vector< void* > inout_buffer_pointers; // buffers for input and output data
+    std::vector< void* > input_buffer_pointers, output_buffer_pointers;
+    for (size_t i = 0; i < engine->getNbBindings(); ++i) {
+        auto binding_size = getSizeByDim(context->getBindingDimensions(i)) * batch_size * sizeof(float);
+        void* buffer_pointer;
+        cudaMalloc(&buffer_pointer, binding_size);
+        inout_buffer_pointers.push_back(buffer_pointer);
+        std::cout << "Allocated " << binding_size << " bytes in GPU, i = " << i << std::endl;
+        if (engine->bindingIsInput(i)) {
+            input_buffer_pointers.push_back(buffer_pointer);
+            input_dims.emplace_back(context->getBindingDimensions(i));
+            std::cout << "Allocated " << binding_size << " bytes in GPU for input" << std::endl;
+        }
+        else {
+            output_buffer_pointers.push_back(buffer_pointer);
+            output_dims.emplace_back(context->getBindingDimensions(i));
+            std::cout << "Allocated " << binding_size << " bytes in GPU for output" << std::endl;
+        }
+    }
+    if (input_dims.empty() || output_dims.empty()) {
+        std::cerr << "Expect at least one input and one output for network" << std::endl;
+        return -1;
+    }
+    if (input_buffer_pointers.size() > 1 || input_buffer_pointers.size() < 1) {
+        std::cerr << "Expect exactly one input for network" << std::endl;
+        return -1;
+    }
+
+    //  Run the test
+    float total1 = 0;
+    float total2 = 0;
+    unsigned miss = 5;
+    for (unsigned i=0; i<filepaths.size(); i++) {
+        const auto t_start1 = std::chrono::high_resolution_clock::now();
+        // preprocess input data
+        cv::Size orig_image_size;
+        if (!processInput(filepaths[i], input_buffer_pointers[0], input_dims[0], image_transformer, orig_image_size)) {
+            std::cerr << "Error while pre-processing image and moving it to GPU device" << std::endl;
+            return -1;
+        }
+        const auto t_start2 = std::chrono::high_resolution_clock::now();
+        // run inference
+        context->enqueueV2(inout_buffer_pointers.data(), 0, nullptr);
```

Comment:
enqueue with a cuda_stream could improve inference performance. cuda_stream could also accelerate cudaMemcpyAsync with pinned memory. Read [this](https://developer.nvidia.com/blog/how-optimize-data-transfers-cuda-cc/) for detail.

Reply:
I've tested hard. pinned memory increases processing speed a bit - I will add it.
cudaMemcpyAsync + enqueueV2 with cuda_stream must be used only within multi-thread wrap. I mean only together with multi-thread functionality implemented. Because I must care myself about write/read data and forward calls queue to prevent starting forward operation before data-to-gpu writing finished and reading detections from the gpu before the forward finished.
Just adding those operators (with '#define CUDA_API_PER_THREAD_DEFAULT_STREAM' of course) doesn't controls operations queue itself. The process runs without crash (thankfully to pinned memory and fixed input layer size) but produces completely incorrect detections.

Reply:
I've tested. Actually this using cuda_streams with async calls doesn't give any speed improvement (organized processing images in 3 cuda streams).
But host memory pinning slightly improved the speed. Will use it.